### PR TITLE
Implementation for code compliance

### PR DESCRIPTION
**Pull request by davidos-b**

### CODE REVIEW

File: include/ardpolyfills/Arduino.hxx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#ifndef SMARTCAR_EMUL_ARDUINO_HXX
+#define SMARTCAR_EMUL_ARDUINO_HXX
+#include <Thread>
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.hxx]: Headers are case-sensitive; it's `<thread>`

File: include/ardpolyfills/Arduino.hxx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#include <chrono>
+#include <iostream>
+#include <stdint.h>
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.hxx]: That's a C header
Use `<cstdint>`

File: include/ardpolyfills/Arduino.hxx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+typedef enum { RIGHTMOSTBIT = 0, LEFTMOSTBIT = 1 } SignificantBit;
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.hxx]: Don't use typedefs; enums can be named directly
```cpp
enum SignificantBit {
  RIGHTMOSTBIT,
  LEFTMOSTBIT
};
```

File: include/ardpolyfills/Arduino.hxx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+typedef enum { HIGH = true, LOW = false } digitalValue;
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.hxx]: Same as above; may use `bool` as underlying type
Also, incorrect naming scheme; types should be PascalCase
```cpp
enum DigitalValue : bool {
  LOW,
  HIGH
};
```

File: include/ardpolyfills/BoardData.hxx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#ifndef SMARTCAR_EMUL_BOARDDATA_HXX
+#define SMARTCAR_EMUL_BOARDDATA_HXX
+#include <atomic>
+#include <range\v3\range\conversion.hpp>
+
+template <std::size_t N>
+
+struct FixBufferBus {
+    std::array<std::byte, N> rx;
+    int len1;
```
Comment (AeroStun) [inline include/ardpolyfills/BoardData.hxx]: Not a descriptive name; suggestion `rx_used_len`

File: include/ardpolyfills/BoardData.hxx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    std::array<std::byte, N> tx;
+    int len2;
```
Comment (AeroStun) [inline include/ardpolyfills/BoardData.hxx]: Same as above

File: include/ardpolyfills/BoardData.hxx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+
+#ifndef SMARTCAR_EMUL_BOARDDATA_HXX
+#define SMARTCAR_EMUL_BOARDDATA_HXX
+#include <atomic>
+#include <range\v3\range\conversion.hpp>
+
+template <std::size_t N>
+
+struct FixBufferBus {
```
Comment (AeroStun) [inline include/ardpolyfills/BoardData.hxx]: Why is there a newline between the template declaration and the class definition?

File: include/ardpolyfills/BoardData.hxx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+#ifndef SMARTCAR_EMUL_BOARDDATA_HXX
+#define SMARTCAR_EMUL_BOARDDATA_HXX
+#include <atomic>
+#include <range\v3\range\conversion.hpp>
```
Comment (AeroStun) [inline include/ardpolyfills/BoardData.hxx]: Unused in current file, therefore should not be present

File: include/ardpolyfills/BoardData.hxx [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    
+struct UartBus : DynaBufferBus {};
+
+struct I2cBus : FixBufferBus<32u> {};
+
+struct BoardData {
+    std::atomic_bool digital_pin_values[32];
+    std::atomic_uint16_t analog_pin_values[32];
+    std::atomic_uint8_t servo_value[10]; // �\_(?)_/� 
+    std::atomic_uint8_t pin_frequency[32];
```
Comment (AeroStun) [inline include/ardpolyfills/BoardData.hxx]: Should be vectors, since different boards have different pin counts
As a general guideline, whenever you have arrays, use at least `std::array` instead of C-style arrays

File: include/ardpolyfills/BoardData.hxx [hunk 10]
```diff
@@ -0,0 +1,52 @@
+/*
+ *  BoardData.hxx
```
Comment (AeroStun) [inline include/ardpolyfills/BoardData.hxx]: The below comments aside, this is good work

File: src/Arduino.cxx [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include <ardpolyfills\Arduino.hxx>
+#include <ardpolyfills\BoardData.hxx>
```
Comment (AeroStun) [inline src/Arduino.cxx]: Non-standard DOS-style paths; use slashes

File: src/Arduino.cxx [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+#include <ardrivo/include/ardrivo/entrypoint.hxx>
```
Comment (AeroStun) [inline src/Arduino.cxx]: You shouldn't have to do this messy include; moving this file to the ardrivo module (where it belongs) will let you have cleaner `#include` directives

File: include/BoardData.hxx [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+#ifndef SMARTCAR_EMUL_BOARDDATA_HXX
+#define SMARTCAR_EMUL_BOARDDATA_HXX
+#include <atomic>
+#include <range\v3\range\conversion.hpp>
```
Comment (AeroStun) [inline include/BoardData.hxx:22]: Unused header in current file; remove include directive
However you use `std::array` without including `<array>`; to be fixed

File: include/BoardData.hxx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+template <std::size_t N>
+
+struct FixBufferBus {
```
Comment (AeroStun) [inline include/BoardData.hxx:26]: Why is there an empty line between the template declaration and the class definition?

File: include/BoardData.hxx [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    std::array<std::byte, N> rx;
+    int len1;
```
Comment (AeroStun) [inline include/BoardData.hxx:28]: Undescriptive name; suggestion: `rx_used_size`

File: include/BoardData.hxx [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    std::array<std::byte, N> tx;
+    int len2;
```
Comment (AeroStun) [inline include/BoardData.hxx:30]: Similar to above

File: include/BoardData.hxx [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    
+struct UartBus : DynaBufferBus {};
+
+struct I2cBus : FixBufferBus<32u> {};
+
+struct BoardData {
+    std::atomic_bool digital_pin_values[32];
+    std::atomic_uint16_t analog_pin_values[32];
+    std::atomic_uint8_t servo_value[10]; // �\_(?)_/� 
+    std::atomic_uint8_t pin_frequency[32];
```
Comment (AeroStun) [inline include/BoardData.hxx:46]: The number of pins is unknown at compile time, since different boards have different pin counts; therefore use `std::vector`

As a general guideline, always use `std::vector`, unless you know the size at compile time, in which case you should use `std::array`

File: include/BoardData.hxx [hunk 18]
```diff
@@ -0,0 +1,52 @@
+/*
+ *  BoardData.hxx
```
Comment (AeroStun) [inline include/BoardData.hxx:2]: The below comments aside, good work

File: include/ardpolyfills/Arduino.h [hunk 19]
```diff
@@ -1,6 +1,6 @@
 /*
- *  Arduino.h
- *  Copyright 2020 ItJustWorksTM
+ *  Arduino.hxx
+ *  Copyright 2020 ItJustWorkTM
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.h:3]: This change should not be present

File: include/ardpolyfills/Arduino.h [hunk 20]
```diff
@@ -16,10 +16,40 @@
  *
  */
 
-#ifndef Arduino_h
-#define Arduino_h
+#ifndef SMARTCAR_EMUL_ARDUINO_HXX
+#define SMARTCAR_EMUL_ARDUINO_HXX
+#include <Thread>
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.h:21]: C++ is case sensitive; include `<thread>` instead

File: include/ardpolyfills/Arduino.h [hunk 21]
```diff
@@ -16,10 +16,40 @@
  *
  */
 
-#ifndef Arduino_h
-#define Arduino_h
+#ifndef SMARTCAR_EMUL_ARDUINO_HXX
+#define SMARTCAR_EMUL_ARDUINO_HXX
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.h:20]: Do not change the header guards name on this file, due to special compatibility with the AVR implementation

File: include/ardpolyfills/Arduino.h [hunk 22]
```diff
@@ -16,10 +16,40 @@
...
+#include <Thread>
+#include <chrono>
+#include <iostream>
+#include <stdint.h>
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.h:24]: `<stdint.h>` is a C header; use `<cstdint>` instead

File: include/ardpolyfills/Arduino.h [hunk 23]
```diff
@@ -16,10 +16,40 @@
...
+
+typedef enum { RIGHTMOSTBIT = 0, LEFTMOSTBIT = 1 } SignificantBit;
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.h:26]: Never use `typedef`. In this case just name the enum:
```cpp
enum SignificantBit {
    RIGHTMOSTBIT,
    LEFTMOSTBIT
};
```

File: include/ardpolyfills/Arduino.h [hunk 24]
```diff
@@ -16,10 +16,40 @@
...
+
+typedef enum { HIGH = true, LOW = false } digitalValue;
```
Comment (AeroStun) [inline include/ardpolyfills/Arduino.h:28]: Same as above; can even use `bool` as underlying type:
Also, incorrect casing: type names should be PascalCase
```cpp
enum DigitalValue : bool {
    LOW,
    HIGH
};
```

File: modules/ardrivo/src/Arduino.cxx [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include <ardpolyfills\Arduino.hxx>
+#include <ardpolyfills\BoardData.hxx>
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:20]: Non-standard DOS-style include paths; could only possibly work on Windows or ReactOS
Also just plain wrong path; should be `#include <Arduino.h>` and `#include <BoardData.hxx>`

File: src/main.cxx [hunk 26]
```diff
@@ -14,5 +14,4 @@
  *  See the License for the specific language governing permissions and
  *  limitations under the License.
  */
-
```
Comment (AeroStun) [inline src/main.cxx:17]: Why this change?

File: modules/ardrivo/src/Arduino.cxx [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+    const auto start_time = std::chrono::steady_clock::now();
+    wait_state_change(state);
+    const auto end = std::chrono::steady_clock::now();
+    return std::chrono::duration_cast<std::chrono::microseconds>(end - start_time).count();
+}
+
+unsigned long pulseInLong(uint8_t pin, uint8_t state, unsigned long timeout) { return pulseIn(pin, state, timeout); }
+
+void shiftOut(uint8_t dataPin, uint8_t clockPin, uint8_t bitOrder, uint8_t val) {
+    uint8_t i;
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:79]: Declare where you use; suggestion: in the init-statement part of the below `for` loop

File: modules/ardrivo/src/Arduino.cxx [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        digitalWrite(clockPin, HIGH);
+        digitalWrite(clockPin, LOW);
+    }
+}
+uint8_t shiftIn(uint8_t dataPin, uint8_t clockPin, uint8_t bitOrder) {
+    uint8_t value = 0;
+    uint8_t hexaDeci = 0x1;
+    uint8_t i;
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:97]: Same as above

File: modules/ardrivo/src/Arduino.cxx [hunk 29]
```diff
@@ -0,0 +1,126 @@
...
+
+    for (i = 0; i < 8; ++i) {
+        digitalWrite(clockPin, hexaDeci);
+        if (bitOrder == 0)
+            value |= digitalRead(dataPin) << i;
+        else
+            value |= digitalRead(dataPin) << (7 - i);
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:104]: Code duplication; refactor so that `value |= digitalRead(dataPin) << ` only appears once

File: modules/ardrivo/src/Arduino.cxx [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+    const auto end = std::chrono::steady_clock::now();
+    return std::chrono::duration_cast<std::chrono::microseconds>(end - start_time).count();
+}
+
+unsigned long pulseInLong(uint8_t pin, uint8_t state, unsigned long timeout) { return pulseIn(pin, state, timeout); }
+
+void shiftOut(uint8_t dataPin, uint8_t clockPin, uint8_t bitOrder, uint8_t val) {
+    uint8_t i;
+
+    for (i = 0; i < 8; i++) {
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:81]: Unused post-increment; please use `++i` as you did line 99

File: modules/ardrivo/src/Arduino.cxx [hunk 31]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+void delay(unsigned long duration) { std::this_thread::sleep_for(std::chrono::milliseconds(duration)); }
+
+void delayMicroseconds(unsigned int duration) { std::this_thread::sleep_for(std::chrono::microseconds(duration)); }
+
+unsigned long micros() {
+
+    auto current_time = std::chrono::steady_clock::now();
+    unsigned long duration = std::chrono::duration_cast<std::chrono::microseconds>(current_time - start_time).count();
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:117]: Can be `const`, therefore should be

File: modules/ardrivo/src/Arduino.cxx [hunk 32]
```diff
@@ -0,0 +1,126 @@
...
+    return duration;
+}
+
+unsigned long millis() {
+
+    auto current_time = std::chrono::steady_clock::now();
+    unsigned long duration = std::chrono::duration_cast<std::chrono::milliseconds>(current_time - start_time).count();
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:124]: Same as above

**Comment by AeroStun** [comment]
Ignore `Outdated` comments; GitHub failed to figure out what to do when you force-pushed while I had a pending review

File: modules/ardrivo/src/Arduino.cxx [hunk 33]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+void tone(uint8_t pin, unsigned int frequency, unsigned long duration) {
+    data.pin_frequency[pin] = frequency;
+    std::this_thread::sleep_for(std::chrono::milliseconds(duration));
+    noTone(pin);
+}
+
+unsigned long pulseIn(uint8_t pin, uint8_t state, unsigned long timeout) {
+    std::atomic_bool& pin_state = data.digital_pin_values[pin];
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:51]: Shouldn't this be a const ref ? (i.e. `const std::atomic_bool&`)

File: modules/ardrivo/src/Arduino.cxx [hunk 34]
```diff
@@ -0,0 +1,126 @@
+/*
+ *  Arduino.cxx
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:2]: Should be added to the sources of ardrivo in ardrivo's CMakeLists.txt

Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:2]: Should be added to the sources of ardrivo in ardrivo's CMakeLists.txt

File: modules/ardrivo/src/Arduino.cxx [hunk 35]
```diff
@@ -0,0 +1,126 @@
...
+int digitalRead(uint8_t pin) { return data.digital_pin_values[pin]; }
+
+void analogWrite(uint16_t pin, uint16_t val) { data.analog_pin_values[pin] = val; }
+
+int analogRead(uint8_t pin) { return data.analog_pin_values[pin]; }
+
+void analogReference(uint8_t mode) {}
+
+void noTone(uint8_t pin) {
+    uint8_t noToneFeq = 255;
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:40]: This should be `constexpr`

File: modules/ardrivo/include/ardrivo/entrypoint.hxx [hunk 36]
```diff
@@ -1,5 +1,5 @@
 /*
- *  entrypoint.hxx
+ *  Entrypoint.hxx
```
Comment (RuthgerD) [inline modules/ardrivo/include/ardrivo/entrypoint.hxx:2]: You forgot to change the filename, CI is failing. (entrypoint -> Entrypoint)
Isn't your IDE warning you about a failed include?

Comment (RuthgerD) [reply modules/ardrivo/include/ardrivo/entrypoint.hxx:2]: Once CI is good this PR is approved for me :+1:

File: modules/ardrivo/src/Arduino.cxx [hunk 37]
```diff
@@ -16,9 +16,9 @@
  *
  */
 
-#include <ardpolyfills\Arduino.hxx>
-#include <ardpolyfills\BoardData.hxx>
-#include <ardrivo/include/ardrivo/entrypoint.hxx>
+#include <BoardData.hxx>
+#include <Arduino.h>
+#include <Entrypoint.hxx>
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:21]: Use quoted include (`#include "file"`) when including project-internal files
(Use angled-bracket include (`#include <file>`) when including system headers)

File: modules/ardrivo/src/entrypoint.cxx [hunk 38]
```diff
@@ -1,5 +1,5 @@
 /*
- *  entrypoint.cxx
+ *  Entrypoint.cxx
```
Comment (AeroStun) [inline modules/ardrivo/src/entrypoint.cxx:2]: Changing the comment doesn't change the actual filename

File: modules/ardrivo/src/Arduino.cxx [hunk 39]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ */
+
+#include <BoardData.hxx>
+#include <Arduino.h>
+#include <Entrypoint.hxx>
+
+using namespace std::literals;
+
+BoardData data;
```
Comment (AeroStun) [inline modules/ardrivo/src/Arduino.cxx:25]: This global variable will be used throughout the entirety of the program; `data` is really generic and context-dependent; `board_data` may be a better fit

Comment (AeroStun) [reply modules/ardrivo/src/Arduino.cxx:25]: While you're at it, `BoardData.hxx` should probably declare this symbol as `extern` too

Comment (AeroStun) [reply modules/ardrivo/src/Arduino.cxx:25]: Oh and make it a pointer, because `init` will initialize it from the runtime

File: modules/ardrivo/include/ardrivo/entrypoint.hxx (hunk 36 above)
Comment (davidos-b) [reply modules/ardrivo/include/ardrivo/entrypoint.hxx:2]: Wellll, nooo
